mkosi/config.py

```diff
@@ -3409,25 +3431,51 @@ def __call__(
                         assert isinstance(v, str)
                         setattr(namespace, s.dest, s.parse(v, getattr(namespace, self.dest, None)))
 
-    def finalize_default(setting: ConfigSetting) -> Optional[Any]:
-        if (v := getattr(namespace, setting.dest, None)) is not None:
-            return v
+    def finalize_value(setting: ConfigSetting) -> Optional[Any]:
+        # If a value was specified on the CLI, it always takes priority. If the setting is a collection of values, we
+        # merge the value from the CLI with the value from the configuration, making sure that the value from the CLI
+        # always takes priority.
+        if (v := getattr(State.cli, setting.dest, setting.parse(None, None))) != setting.parse(None, None):
+            if isinstance(v, list):
+                return getattr(State.config, setting.dest, []) + v
+            elif isinstance(v, tuple):
+                return (*getattr(State.config, setting.dest, tuple()), *v)
+            elif isinstance(v, dict):
+                return getattr(State.config, setting.dest, {}) | v
+            elif isinstance(v, set):
+                return getattr(State.config, setting.dest, set()) | v
+            else:
+                return v
 
-        for d in setting.default_factory_depends:
-            finalize_default(SETTINGS_LOOKUP_BY_DEST[d])
+        # If the setting was assigned the empty string on the CLI, we don't use any value configured in the
+        # configuration file. Additionally, if the setting is a collection of values, we won't use any default
+        # value either if the setting is set to the empty string on the command line.
 
-        # If the setting was assigned the empty string, we don't use any configured default value.
-        if not hasattr(namespace, setting.dest) and setting.dest in defaults:
-            default = getattr(defaults, setting.dest)
-        elif setting.default_factory:
-            default = setting.default_factory(namespace)
-        elif setting.default is None:
+        if not hasattr(State.cli, setting.dest) and (v := getattr(State.config, setting.dest, None)) is not None:
+            return v
+
+        if hasattr(State.cli, setting.dest) and isinstance(setting.parse(None, None), (dict, tuple, list, set)):
             default = setting.parse(None, None)
-        else:
+        elif setting.default_factory:
+            # To determine default values, we need the final values of various settings in
+            # a namespace object, but we don't want to copy the final values into the config
+            # namespace object just yet so we create a new namespace object instead.
+            factoryns = argparse.Namespace(**{
+                d: finalize_value(SETTINGS_LOOKUP_BY_DEST[d]) for d in setting.default_factory_depends
+            })
```

Comment:
I think this way it's easier to not overlook that unpacking is happening.
```suggestion
            factoryns = argparse.Namespace(
                **{d: finalize_value(SETTINGS_LOOKUP_BY_DEST[d]) for d in setting.default_factory_depends}
            )
```